src/mlpack/core/data/split_data.hpp

```diff
@@ -224,6 +229,102 @@ Split(const arma::Mat<T>& input,
                          std::move(testData));
 }
 
+/**
+ * Performs train test split.
+ *
+ * @tparam DatasetType Type of input dataset.
+ * @tparam LabelsType Type of input labels.
+ *
+ * @param dataset Features of dataset.
+ * @param labels Labels of the dataset.
+ * @param validRatio Ratio for train-test split.
+ * @param shuffle Boolean to determine shuffling of dataset.
+ */
+template<typename DatasetX>
+void Split(DatasetX& dataset,
+                    std::deque<arma::vec>& labels,
+                    DatasetX& trainFeatures,
+                    arma::field<arma::vec>& trainLabels,
+                    DatasetX& validFeatures ,
+                    arma::field<arma::vec>& validLabels,
+                    const double validRatio,
+                    const bool shuffle)
+{
+  const size_t validSize = static_cast<size_t>(dataset.n_cols * validRatio);
+  const size_t trainSize = dataset.n_cols - validSize;
+
+  arma::uvec order = arma::linspace<arma::uvec>(0, dataset.n_cols - 1,
+      dataset.n_cols);
+  if (shuffle)
+    order = arma::shuffle(order);
+
+  if (trainSize > 0)
+  {
+    trainLabels.set_size(1, trainSize);
+    trainFeatures = dataset.cols(order.subvec(0, trainSize - 1));
+
+    // Field type has fixed size so we can't use span and assignment
+    // operator.
+    for (size_t i = 0; i < trainSize; i++)
+      trainLabels(0, i) = labels[i];
+  }
+
+  if (validSize <= dataset.n_cols)
+  {
+    validFeatures = dataset.cols(order.subvec(trainSize,
+        dataset.n_cols - 1));
```

Comment:
Asume that your input will also be of field type and implement split in field type. I used deque for simplicity while parsing XML but a user in general would call it on field type.

Reply:
That goes for both input and inputLabels I guess, so I'll remove DatasetX/Y template?

Reply:
Yes. And DatasetX is more suited for models than mlpack. The input / inputLabels you used are better.